File: server/service-notes/src/notes_service.rs
use anyhow::Context;
use anyhow::Result;
use futures::TryStreamExt;
use sqlx::types::time::OffsetDateTime;
use sqlx::{postgres::PgRow, query, types::Uuid, Row};
use tokio::sync::mpsc;
use tokio_stream::wrappers::ReceiverStream;
use tonic::{Request, Response, Status};

use crate::{
    proto::{notes_service_server::NotesService, Note, NoteId, UserId},
    MyNotes,
};

fn map_note(row: Option<PgRow>) -> Result<Note> {
    match row {
        Some(row) => {
            let id: Uuid = row.try_get("id").context("Failed to get id")?;
            let user_id: Uuid = row.try_get("userId").context("Failed to get userId")?;
            let created: OffsetDateTime =
                row.try_get("created").context("Failed to get created")?;
            let updated: OffsetDateTime =
                row.try_get("updated").context("Failed to get updated")?;
            let deleted: Option<OffsetDateTime> =
                row.try_get("deleted").context("Failed to get deleted")?;
            let note = Note {
                id: id.to_string(),
                user_id: user_id.to_string(),
                title: row.try_get("title").context("Failed to get title")?,
                content: row.try_get("content").context("Failed to get content")?,
                created: created.to_string(),
                updated: updated.to_string(),
                deleted: deleted.map(|d| d.to_string()),
                user: None,
            };
            Ok(note)
        }
        None => Err(anyhow::anyhow!("Note not found")),
    }
}

#[tonic::async_trait]
impl NotesService for MyNotes {
    type GetNotesStream = ReceiverStream<Result<Note, Status>>;

    async fn get_notes(
        &self,
        request: Request<UserId>,
    ) -> Result<Response<Self::GetNotesStream>, Status> {
        println!("GetNotes = {:?}", request);
        let start = std::time::Instant::now();

        let self_pool = self.pool.clone();
        let (tx, rx) = mpsc::channel(4);
        let user_id = request.into_inner().user_id;
        let uuid = Uuid::parse_str(&user_id).map_err(|e| Status::internal(e.to_string()))?;

        tokio::spawn(async move {
            let mut notes_stream = query("SELECT * FROM notes WHERE \"userId\" = $1 and deleted is null order by created desc")
                .bind(uuid)
                .fetch(&self_pool);

            loop {
                match notes_stream.try_next().await {
                    Ok(None) => {
                        let elapsed = start.elapsed();
                        println!("Elapsed: {:.2?}", elapsed);
                        break;
                    }
                    Ok(note) => {
                        let note = map_note(note).unwrap();
                        tx.send(Ok(note)).await.unwrap();
                    }
                    Err(e) => {
                        tx.send(Err(Status::internal(e.to_string()))).await.unwrap();
                        break;
                    }
                }
            }
        });
        Ok(Response::new(ReceiverStream::new(rx)))
    }

    async fn create_note(&self, request: Request<Note>) -> Result<Response<Note>, Status> {
        println!("CreateNote = {:?}", request);
        let start = std::time::Instant::now();

        let note = request.into_inner();
        let user_id =
            Uuid::parse_str(&note.user_id).map_err(|e| Status::internal(e.to_string()))?;
        let self_pool = self.pool.clone();

        let row =
            query("INSERT INTO notes (title, content, \"userId\") VALUES ($1, $2, $3) RETURNING *")
                .bind(note.title)
                .bind(note.content)
                .bind(user_id)
                .fetch_one(&self_pool)
                .await
                .map_err(|e| Status::internal(e.to_string()))?;

        let note = map_note(Some(row)).map_err(|e| Status::internal(e.to_string()))?;
        let elapsed = start.elapsed();
        println!("Elapsed: {:.2?}", elapsed);
        return Ok(Response::new(note));
    }
    async fn delete_note(&self, request: Request<NoteId>) -> Result<Response<Note>, Status> {
        println!("DeleteNote = {:?}", request);
        let start = std::time::Instant::now();

        let self_pool = self.pool.clone();

        let request = request.into_inner();
        let note_uuid =
            Uuid::parse_str(&request.note_id).map_err(|e| Status::internal(e.to_string()))?;
        let user_uuid =
            Uuid::parse_str(&request.user_id).map_err(|e| Status::internal(e.to_string()))?;

        let row =
            query("UPDATE notes SET deleted = NOW() WHERE id = $1 AND \"userId\" = $2 RETURNING *")
                .bind(note_uuid)
                .bind(user_uuid)
                .fetch_one(&self_pool)
                .await
                .map_err(|e| Status::not_found(e.to_string()))?;

        let note = map_note(Some(row)).map_err(|e| Status::internal(e.to_string()))?;
        let elapsed = start.elapsed();
        println!("Elapsed: {:.2?}", elapsed);
        return Ok(Response::new(note));
    }
}

File: service-notes-go/src/notes_service.go
package main

import (
	"context"
	"database/sql"
	"log/slog"
	"time"

	pb "rusve/proto"

	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

func (s *server) GetNotes(in *pb.UserId, stream pb.NotesService_GetNotesServer) error {
	start := time.Now()

	rows, err := db.Query(`select * from notes where user_id = $1 and deleted is null`, in.UserId)
	if err != nil {
		slog.Error("db.Query", "error", err)
		return err
	}
	defer rows.Close()

	for rows.Next() {
		note, err := mapNote(rows, nil)
		if err != nil {
			slog.Error("mapNote", "error", err)
			return err
		}
		err = stream.Send(note)
		if err != nil {
			slog.Error("stream.Send", "error", err)
			return err
		}
	}
	if rows.Err() != nil {
		slog.Error("rows.Err", "error", err)
		return err
	}
	slog.Info("GetNotes", "time", time.Since(start))
	return nil
}

func (s *server) CreateNote(ctx context.Context, in *pb.Note) (*pb.Note, error) {
	start := time.Now()
	rules := map[string]string{
		"UserId":  "required,max=100",
		"Title":   "required,max=100",
		"Content": "required,max=1000",
	}
	validate.RegisterStructValidationMapRules(rules, pb.Note{})
	err := validate.Struct(in)
	if err != nil {
		slog.Error("validate.Struct", "error", err)
		return nil, status.Error(codes.InvalidArgument, "Invalid argument")
	}

	var row *sql.Row
	if in.Id == "" {
		row = db.QueryRow(`insert into notes (user_id, title, content) values ($1, $2, $3) returning *`, in.UserId, in.Title, in.Content)
	} else {
		row = db.QueryRow(`update notes set title = $1, content = $2 where id = $3 and user_id = $4 returning *`, in.Title, in.Content, in.Id, in.UserId)
	}
	note, err := mapNote(nil, row)
	if err != nil {
		slog.Error("mapNote", "error", err)
		return nil, err
	}

	slog.Info("CreateNote", "time", time.Since(start))
	return note, nil
}

func (s *server) DeleteNote(ctx context.Context, in *pb.NoteId) (*pb.Empty, error) {
	start := time.Now()

	_, err := db.Exec(`update notes set deleted = now() where id = $1 and user_id = $2`, in.NoteId, in.UserId)
	if err != nil {
		slog.Error("db.Exec", "error", err)
		return nil, err
	}

	slog.Info("DeleteNote", "time", time.Since(start))
	return &pb.Empty{}, nil
}
